include/osquery/registry.h

```diff
@@ -15,14 +15,294 @@
 #include <unordered_map>
 #include <utility>
 
+#include <functional>
+#include <map>
+#include <mutex>
+#include <vector>
+
+#include <boost/noncopyable.hpp>
+
 #include <osquery/database/results.h>
 #include <osquery/logger.h>
 
-#include "osquery/registry/init_registry.h"
-#include "osquery/registry/singleton.h"
-
 namespace osquery {
 
+typedef std::map<std::string, std::string> RouteInfo;
+typedef std::map<std::string, RouteInfo> RegistryRoutes;
+typedef std::map<std::string, RegistryRoutes> RegistryBroadcast;
+
+typedef std::map<std::string, std::string> PluginRequest;
+typedef std::map<std::string, std::string> PluginReponse;
+
+template <class RegistryType>
+class NewRegistry {
+ protected:
+  typedef std::shared_ptr<RegistryType> RegistryTypeRef;
+  virtual void type() {}
+
+ public:
+  template <class Item>
+  Status add(const std::string& item_name) {
+    if (items_.count(item_name) > 0) {
+      return Status(1, "Duplicate registry item exists: " + item_name);
+    }
+    auto item = reinterpret_cast<RegistryType*>(new Item());
+
+    auto status = item->setUp();
+    if (!status.ok()) {
+      return status;
```

Comment:
This looks like a memory leak here. Can we just create the `shared_ptr` on line 50 instead of 57?